plugins/liquid/refined-types-plugin/src/main/kotlin/arrow/meta/plugins/liquid/phases/analysis/solver/ConstraintChecks.kt

```diff
@@ -350,17 +392,42 @@ private fun SolverState.checkCallExpression(
 private fun SolverState.checkCallArguments(
   resolvedCall: ResolvedCall<out CallableDescriptor>,
   data: CheckData
-): ContSeq<List<Pair<String, String>>> =
-  resolvedCall.allArgumentExpressions().map { (name, _, expr) ->
-    val argUniqueName =
-      if (expr != null && solver.isResultReference(expr, data.context.trace.bindingContext)) {
-        RESULT_VAR_NAME
-      } else {
-        names.newName(name)
+): ContSeq<Either<ExplicitReturn, List<Pair<String, String>>>> {
+  // why is this so complicated?
+  //   in theory we just need to run checkExpressionConstraints over each argument
+  //   (in fact, the original implementation just did that, and then called .sequence())
+  //   alas, Kotlin allows arguments to include 'return' (yes, really!)
+  //   so we need to check after each step whether a ExplicitReturn has been generated
+  //   and in that case we stop the check of any more arguments
+  //   (stopping there is important, since introducing additional constraints from
+  //    other arguments may not be right in the general case)
+  fun <A> go(expressions: List<Triple<String, A, KtExpression?>>): ContSeq<Either<ExplicitReturn, List<Pair<String, String>>>> =
```

Comment:
These can typically also be implemented using `scan`, or `runningFold`. There is a large set of these functions scan/reduce/fold but they all have different names than you'd find them in Haskell.

[runningFold](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/running-fold.html)
[scan](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/scan.html)
[collection ops](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections)

Reply:
Thanks for the pointers. At the end, it turned out that a simple `fold` was enough, with the caveat that now if an argument sets an `ExplicitReturn`, it's still threaded through the rest of the arguments (although unchanged). I don't think this will have any performance implication, since argument lists are short and the explicit return case almost never happens.

Reply:
[Here's the diff](https://github.com/arrow-kt/arrow-meta/pull/845/commits/bdb168184f10254af06ecf4a6b0e238b1e67f631) in case you want to have a closer look.